dask/layers.py

```diff
@@ -844,3 +868,85 @@ def _construct_graph(self, deserializing=False):
             dsk[(self.name, i)] = (concat_func, _concat_list)
 
         return dsk
+
+
+class DataFrameIOLayer(Blockwise, DataFrameLayer):
+    """DataFrame-based Blockwise Layer with IO
+
+    Parameters
+    ----------
+    name : string
+        Name to use for the constructed layer.
+    columns : string, list or None
+        Field name(s) to read in as columns in the output.
+    inputs : list[tuple]
+        List of input-argument tuples. Each element will be
+        passed to ``io_func`` to construct a single output
+        partition.
+    io_func : callable
+        A callable function that takes in a single tuple
+        of arguments, and outputs a DataFrame partition.
+    """
+
+    def __init__(
+        self,
+        name,
+        columns,
+        inputs,
+        io_func,
+        part_ids=None,
+        label=None,
+        produces_tasks=False,
+        annotations=None,
+    ):
+        self.name = name
+        self.columns = columns
+        self.inputs = inputs
+        self.io_func = io_func
+        self.part_ids = list(range(len(inputs))) if part_ids is None else part_ids
+        self.label = label
+        self.produces_tasks = produces_tasks
+        self.annotations = annotations
+
+        # Define mapping between key index and "part"
+        io_arg_map = BlockwiseDepDict(
+            {(i,): self.inputs[i] for i in self.part_ids},
+            produces_tasks=self.produces_tasks,
+        )
+
+        # Use Blockwise initializer
+        dsk = {self.name: (io_func, blockwise_token(0))}
+        super().__init__(
+            output=self.name,
+            output_indices="i",
+            dsk=dsk,
+            indices=[(io_arg_map, "i")],
+            numblocks={},
+            annotations=annotations,
+        )
+
+    def project_columns(self, columns):
+        # Method inherited from `DataFrameLayer.project_columns`
+        if columns and (self.columns is None or columns < set(self.columns)):
+            layer = DataFrameIOLayer(
+                (self.label or "subset-") + tokenize(self.name, columns),
+                list(columns),
+                self.inputs,
+                self.io_func,
+                part_ids=self.part_ids,
+                produces_tasks=self.produces_tasks,
+                annotations=self.annotations,
+            )
+            if hasattr(layer.io_func, "columns"):
+                # Apply column projection
+                layer.io_func = copy.deepcopy(layer.io_func)
+                layer.io_func.columns = columns
```

Comment:
nit: probably overkill, but maybe having a `project_columns` method on `DataFrameIOLayer` would be nice, just to make it clearer that such a thing can be done, and how it works? Whereas this feels like reaching into/making assumptions about the implementation of a `DataFrameIOLayer`.

Reply:
I don't think I am understanding the request.  This code is in `DataFrameIOLayer.project_columns`. Is that what you are asking for?

Reply:
Sorry, I very much mis-wrote that. What I should have said was, maybe the `io_func`s should implement a `project_columns` method? Like on `ParquetFunctionWrapper`, `CsvFunctionWrapper`, etc.? So this could look like
```python
try:
    layer.io_func = layer.io_func.project_columns(columns)
except AttributeError:
    pass
```

Reply:
okay, got it. Makes sense to me